### ISSUE TITLE
generic inference fails with function argument type like `Foo[T, object] | Foo[object, T]`

### ISSUE THREAD

**Issue by DetachHead** [astral-sh/ty#3533]
### Summary

`main.py`:

```py
class Foo[T, U]:
    def t(self) -> T: ...
    def u(self) -> U: ...


def fn[T](foo: Foo[T, object] | Foo[object, T], value: T) -> T: ...


def fn2[T](value: T) -> T:
    return fn(Foo[T, object](), value) # Object of type `object` is not assignable to `T@fn2` (invalid-assignment)
```

[Playground](https://play.ty.dev/ca1a0b33-b116-4bff-afb8-4c892486b209)

this works correctly in [pyright](https://basedpyright.com/?typeCheckingMode=all&reportUnusedParameter=false&reportUnusedVariable=false&code=MYGwhgzhAEBiD28DaAVANNAqgXQFwFgAoaE6AEwFMAzaAFwAoIKQqBKaAWgD5oVdoAdEKKly1aAFdGzNpx6Z%2BQgURWFKNKgDtU2elUT8EydNHgAjAFYVgtbNAA%2BcREnNWbGFNgwA3MCAkU-Cjs3LyKwoSq6tBaAEw69L7%2BgbwhPHwipABOFLQSWZoxmvRGqBiu1rb0rD5%2BAaxAA) and [pyrefly](https://pyrefly.org/sandbox/?project=N4IgZglgNgpgziAXKOBDAdgEwEYHsAeAdAA4CeSIAxlKnHAAQBiuuA2gCoA09AqgLqIAOunqj6mGGHoAXABRwYUMAEp6AWgB89donqF9wseMn0ArvMUr1Wnrv2FhjrCbDoOfWWBa7mbLvVxsACsYSmk%2BegAfJhZWQJCw7nY%2BbgA3VChTGF12VU1tOwN0JwkpVwAmd1l0zOztPK0dQzEAJxhpUxaRV1lfDm540PDZZTSMrOUQThAyNrAoUkJpXABbKAoyaQALXHQ1VJgWuAhd%2BgBeekEQAGZCAEZyq%2BFWQ5bcI75hDHRcaVRpE57FYQOArf6ULbnegAcle72hwhAAF9pqgwhADoxoDAKGgsHgiGRkUA) (although pyrefly has a very similar issue: https://github.com/facebook/pyrefly/issues/3583)

### Version

73de9b45c

**Comment by dhruvmanila** [comment]
I see that the diagnostic raised is actually `invalid-return-type` instead of `invalid-assignment`.

And, the error goes away if I remove the `Foo[object, T]` from the union of `foo` annotation.

This might be a limitation of the old constraint solver and is related to #2870 and #2799. The `Foo[T, object]` can match both the union element and we resolve the return type accounting for both the possibilities which is why you're seeing `object` as the return type of `foo` call.

**Comment by dhruvmanila** [comment]
Looking at #2870 more, I think it's a duplicate of that.
